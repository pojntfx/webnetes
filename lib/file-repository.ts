import WebTorrent from "webtorrent-hybrid";
import { getLogger } from "./logger";

export class FileRepository {
  private logger = getLogger();

  private client?: WebTorrent.Instance;

  async open() {
    this.logger.verbose("Opening file repository");

    this.client = new WebTorrent();
  }

  async close() {
    this.logger.verbose("Closing file repository");

    await new Promise<void>((res, rej) =>
      this.client?.destroy((e: any) => (e ? rej(e) : res()))
    );
  }

  async seed(content: Uint8Array) {
    return "";
  }

  async add(magnetURI: string) {
    return new Uint8Array();
  }

  async remove(magnetURI: string) {}
}
