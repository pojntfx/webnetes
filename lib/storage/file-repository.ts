import WebTorrent from "webtorrent-hybrid";
import { ClosedError } from "../errors/closed";
import { FileNotInTorrentError } from "../errors/file-not-in-torrent";
import { getLogger } from "../utils/logger";

const FILE_NAME = "data";

export class FileRepository {
  private logger = getLogger();

  private client?: WebTorrent.Instance;

  async open() {
    this.logger.verbose("Opening file repository");

    this.client = new WebTorrent();
  }

  async close() {
    this.logger.verbose("Closing file repository");

    await new Promise<void>((res, rej) =>
      this.client?.destroy((e: any) => (e ? rej(e) : res()))
    );
  }

  async seed(content: Uint8Array) {
    if (this.client) {
      return await new Promise<string>((res, rej) => {
        try {
          const dataToSeed = Buffer.from(content);

          (dataToSeed as any)["name"] = FILE_NAME;

          this.client!.seed(dataToSeed, {}, (
            torrent // We check above
          ) => res(torrent.magnetURI));
        } catch (e) {
          rej(e);
        }
      });
    } else {
      throw new ClosedError("FileRepository");
    }
  }

  async add(magnetURI: string) {
    if (this.client) {
      return await new Promise<Uint8Array>((res, rej) => {
        try {
          this.client!.add(magnetURI, {}, (torrent) => {
            // We check above
            const file = torrent.files.find((t) => t.name === FILE_NAME);

            if (file) {
              file!.getBuffer((e, buf) => {
                if (e) {
                  rej(e);
                } else {
                  res(new Uint8Array(buf!)); // We check with e above
                }
              }); // We check above
            } else {
              rej(new FileNotInTorrentError());
            }
          });
        } catch (e) {
          rej(e);
        }
      });
    } else {
      throw new ClosedError("FileRepository");
    }
  }

  async remove(magnetURI: string) {
    if (this.client) {
      await new Promise<void>((res, rej) =>
        this.client!.remove(magnetURI, {}, (e) => (e ? rej(e) : res()))
      ); // We check above
    } else {
      throw new ClosedError("FileRepository");
    }
  }
}
