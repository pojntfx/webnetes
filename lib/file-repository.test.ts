import { FileRepository } from "./file-repository";

describe("FileRepository", () => {
  let fileRepository: FileRepository;

  beforeEach(() => {
    fileRepository = new FileRepository();
  });

  describe("lifecycle", () => {
    describe("open", () => {
      it("should open", async () => {
        await fileRepository.open();
      });

      afterEach(async () => {
        await fileRepository.close();
      });
    });

    describe("close", () => {
      beforeEach(async () => {
        await fileRepository.open();
      });

      it("should close", async () => {
        await fileRepository.close();
      });
    });
  });

  describe("non-lifecycle", () => {
    describe("seed", () => {
      beforeEach(async () => {
        await fileRepository.open();
      });

      afterEach(async () => {
        await fileRepository.close();
      });

      describe("seeding", () => {
        it("should seed an empty Uint8Array", async () => {
          const magnetURI = await fileRepository.seed(new Uint8Array());

          expect(magnetURI).toMatchSnapshot();
        }, 20000);
      });

      describe("adding", () => {
        const inFile = new TextEncoder().encode("Hello, world!");
        let magnetURI = "";
        let seedingRepo: FileRepository;

        beforeAll(async () => {
          seedingRepo = new FileRepository();

          await seedingRepo.open();

          magnetURI = await seedingRepo.seed(inFile);
        }, 20000);

        afterAll(async () => {
          await seedingRepo.close();
        });

        it("should be able to add a file", async () => {
          const outFile = await fileRepository.add(magnetURI);

          expect(outFile).toStrictEqual(inFile);
        }, 20000);
      });
    });
  });
});
