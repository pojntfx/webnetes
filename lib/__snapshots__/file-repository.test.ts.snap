// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileRepository non-lifecycle seed seeding should seed an empty Uint8Array 1`] = `"magnet:?xt=urn:btih:cf117d0f40dfe297f779602fe3a11d004b5509cf&dn=data&tr=wss%3A%2F%2Ftracker.btorrent.xyz&tr=wss%3A%2F%2Ftracker.openwebtorrent.com&tr=udp%3A%2F%2Ftracker.leechers-paradise.org%3A6969&tr=udp%3A%2F%2Ftracker.coppersurfer.tk%3A6969&tr=udp%3A%2F%2Ftracker.opentrackr.org%3A1337&tr=udp%3A%2F%2Fexplodie.org%3A6969&tr=udp%3A%2F%2Ftracker.empire-js.us%3A1337"`;
